### PR TITLE
[Feature] Migrate database to and from encryption at rest

### PR DESCRIPTION
**Pull request by johnxnguyen**
## What's new in this PR?

When enabling encryption at rest, we now migrate existing (relevant) data in the database so that it is in the same state as it would be if it was created while the feature was already enabled. Conversely, when we disable encryption at rest, we migrate the same data back to the normal unencrypted state.

The relevant entities are:

- `ZMGenericMessageData`: contains the message content which we encrypt / decrypt when migrating.
- `ZMMessage`: an abstract parent class that contains references to `ZMGenericMessageData`. We need to update / clear the `normalizedText` property, which contains message content in plaintext optimized for search queries.
- `ZMConversation`: it may have a draft message, which is a message that the user has typed but not yet sent. We need to encrypt / decrypt this message when migrating.

We also introduce the `migrateLocalStorage()` method on the `StorageStack`, which performs changes to a local store in an atomic manner by operating on a temporary copy of the local store. This method is used in `wire-ios-sync-engine` to perform the encryption at rest migration in a safe and performant way.

### CODE REVIEW

**Comment by typfel** [comment]
Any idea what the perforis like? Have you tried this out on an account with a large history?

**Comment by johnxnguyen** [comment]
> Any idea what the performance is like? Have you tried this out on an account with a large history?

@typfel not yet but I'll try it out today.

File: Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift [hunk 1]
```diff
@@ -19,29 +19,157 @@
 import Foundation
 import WireCryptobox
 
-extension NSManagedObjectContext {
+extension Sequence where Element: NSManagedObject {
     
-    enum EncryptionError: LocalizedError {
+    /// Perform changes on a sequence of NSManagedObjects and save at a regular internal and fault
```
Comment (johnxnguyen) [inline Source/ManagedObjectContext/NSManagedObjectContext+EncryptionAtRest.swift:24]:
```suggestion
    /// Perform changes on a sequence of NSManagedObjects and save at a regular interval and fault
```

File: Tests/Source/ManagedObjectContext/NSManagedObjectContextTests+EncryptionAtRest.swift [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        let request = NSFetchRequest<T>(entityName: T.entityName())
+        request.returnsObjectsAsFaults = false
+        return try request.execute()
+    }
+
+    // MARK: - Positive Tests
+
+    // MARK: - Message Content
+
+    func test_existing_message_content_is_encrypted_when_ear_is_enabled() throws {
```
Comment (johnxnguyen) [inline Tests/Source/ManagedObjectContext/NSManagedObjectContextTests+EncryptionAtRest.swift:43]: I suppose we should change the test names to camel case.

**Review by johnxnguyen — COMMENTED** [review]
Looks good to me. @marcomaddy1982 you assigned me but I can't approve because I was the one that opened this PR.

File: Tests/Source/ManagedObjectContext/NSManagedObjectContextTests+EncryptionAtRest.swift (hunk 2 above)
Comment (typfel) [reply Tests/Source/ManagedObjectContext/NSManagedObjectContextTests+EncryptionAtRest.swift]: Done

**Review by marcomaddy1982 — APPROVED** [review]

**Review by typfel — APPROVED** [review]

**Pull request merged**